File: Programmers/[60058] 괄호 변환/woong-jae/60058.js
function solution(p) {
    const isCorrect = p => {
        let count = 0;
        for(const bracket of p) {
            if(bracket === "(") count++;
            else count--;
            if(count < 0) return false;
        }
        return true;
    }

    const getCorrectBracketString = p => {
        if(!p) return "";

        let index = 0, bracketCount = { left: 0, right: 0 };
        while(index < p.length) {
            const bracket = p[index];
            bracketCount[bracket === "(" ? "left" : "right"] += 1;
            index++;
            if(bracketCount.left === bracketCount.right) break;
        }
        const u = p.slice(0, index), v = p.slice(index);

        if(isCorrect(u)) return u + getCorrectBracketString(v);
        return "(" + getCorrectBracketString(v) + ")" + u
            .slice(1, index - 1)
            .split("")
            .map(char => char === ")" ? "(" : ")")
            .join("");
    }

    return getCorrectBracketString(p);
}

File: Programmers/[60058] 괄호 변환/Johoseong/60058.py
def is_correct(p):
    stack = []
    for i in p:
        if i == '(': 
            stack.append(i)
        else:
            if not stack: return False 
            stack.pop()
    return True
    
def convert(u):
    tmp = ''
    for l in u:
        if l == '(': tmp += ')'
        else: tmp += '('
    return tmp

def divide(p):
    left, right = 0, 0
    for i in p:
        if i == '(': left += 1
        else: right += 1
        if left == right: # 괄호 세트 찾았으면 멈춤
            break
    return p[:(left + right)], p[(left + right):]

def solution(p):
    answer = ''
    while len(p) != 0:
        u, p = divide(p)
        if is_correct(u):
            answer += u
        else:
            answer += ('(' + solution(p) + ')' + convert(u[1:-1]))
            break
    return answer

print(solution("(()())()"))